netty/src/main/java/io/grpc/netty/WriteQueue.java

```diff
@@ -120,15 +133,25 @@ ChannelFuture enqueue(Object command, ChannelPromise promise, boolean flush) {
   private void flush() {
     try {
       boolean flushed = false;
+      // We can't just call flush after having completely drained the queue, as new objects might
+      // be added to the queue continuously and so it may never get empty. If we never flushed in
+      // that case we would be guaranteed to OOM.
+      // However, we also don't want to flush too often as flushing invokes expensive system
+      // calls and we want to feed Netty with enough data so as to maximize the data written per
+      // system call.
+      int writesBeforeFlush = min(queue.size(), MAX_WRITES_BEFORE_FLUSH);
+      // Always dequeue in chunks, in order to not acquire the queue's lock too often.
       while (queue.drainTo(writeChunk, DEQUE_CHUNK_SIZE) > 0) {
-        while (writeChunk.size() > 0) {
-          writeChunk.poll().run();
+        writesBeforeFlush -= writeChunk.size();
+        while (!writeChunk.isEmpty()) {
+          QueuedCommand cmd = writeChunk.poll();
+          channel.write(cmd.command(), cmd.promise());
+        }
+        if (writesBeforeFlush <= 0) {
```

Comment:
@nmittler The condition `if (!flushed)` is only for when there are zero items in the queue. The last items also get flushed cause `writesBeforeFlush` is effectively equal to `queue.size()`. So it's called after having written `queue.size()` many items, which we will always be doing. That being said, I think the code is more complex than it needs to be. I ll try to make it more clear.